concepts/pipeline/http_edge_spec.md

```diff
@@ -0,0 +1,105 @@
+# HTTP Edge Server Specification
+
+This document specifies the behavior of the server which is accepts Telemetry submissions from Firefox.
+
+The original implementation of the HTTP Edge Server was tracked in [Bug 1129222](https://bugzilla.mozilla.org/show_bug.cgi?id=1129222).
+
+## General Data Flow
+HTTP submissions come in from the wild, hit a load balancer, then optionally an nginx proxy, then the HTTP Edge Server described in this document. Data is accepted via a POST/PUT request from clients, which the server will wrap in a [Heka message](https://hekad.readthedocs.io/en/dev/message/index.html) and forward to two places: the [Services Data Pipeline](data_pipeline.md), where any further processing, analysis, and storage will be handled; as well as to a short-lived S3 bucket which will act as a fail-safe in case there is a processing error and/or data loss within the main Data Pipeline.
+
+## GET Requests
+Accept GET on `/status`, returning `OK` if all is well. This can be used to check the health of web servers.
+
+## POST/PUT Requests
+Treat POST and PUT the same. Accept POST or PUT to URLs of the form
+
+`^/submit/namespace/[id[/dimensions]]$`
+
+Example Telemetry format:
+
+`/submit/telemetry/docId/docType/appName/appVersion/appUpdateChannel/appBuildID`
+
+Specific Telemetry example:
+
+`/submit/telemetry/ce39b608-f595-4c69-b6a6-f7a436604648/main/Firefox/61.0a1/nightly/20180328030202`
+
+Note that `id` above is a unique document ID, which is used for de-duping submissions. This is *not* intended to be the `clientId` field from Telemetry. If `id` is omitted, we will not be able to de-dupe based on submission URLs.
+
+### Namespaces
+Each allowed namespace should have some configuration, and it should be relatively easy to add new namespaces.
+
+Per-namespace configuration:
+* Required: Max Payload size
+* Required: Max path length
+* Required: Include Client IP y/n
+* Optional: Mapping of dimension index -> field name
+* Optional: “logger” value (default is to use the namespace, but could be useful to re-map)
+
+### Creating a Heka message from submission
+Main message fields:
+* POST Body -> `Payload`
+* Client IP -> `remote_addr` (if indicated by the namespace config)
+* Now -> `timestamp` ("Now" means the current time on the edge server node. Should use `ntp` or similar to ensure correct server time)
+* namespace -> `logger`
+* id -> `uuid` (if supplied, otherwise randomly generate a UUID)
+* Edge hostname -> `hostname`
+* Type -> `incoming` (open to suggestions for naming) Some indication that this is a raw message.
+
+Other fields:
+* Store the part of the path after the namespace in `Fields["path"]`
```

Comment:
There is no Fields[path] entry.  The additional fields follow:

# Additional Fields
- Fields[uri] == "/submit/telemetry/28c59e5a-8efe-4144-b43c-3d82846d93bb/main/Firefox/52.6.0/esr/20180118122319"
- Fields[args] == nil or the provided query string (e.g. "v=4")
- Fields[protocol] == "HTTP/1.1"

## Configuration requested header fields
Telemetry specific configuration: https://github.com/mozilla-services/puppet-config/blob/06ec0beae535184cc7455c3c2c32b6571160196d/pipeline/modules/pipeline/templates/edge/nginx.conf.erb#L31
- Fields[Content-Length] == "4722"
- Fields[Date] == "Mon, 12 Mar 2018 00:02:18 GMT"
- Fields[DNT] == "1"
- Fields[User-Agent] == "Mozilla/5.0 (Windows NT 5.1; rv:52.0) Gecko/20100101 Firefox/52.0"
- Fields[X-Forwarded-For] == "127.0.0.1"
- Fields[Host] == "incoming.telemetry.mozilla.org"
- Fields[X-PingSender-Version] == "1.0"